packages/utils/src/instrument/xhr.ts

```diff
@@ -0,0 +1,158 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+/* eslint-disable @typescript-eslint/ban-types */
+import type { HandlerDataXhr, SentryWrappedXMLHttpRequest, WrappedFunction } from '@sentry/types';
+
+import { isString } from '../is';
+import { fill } from '../object';
+import { GLOBAL_OBJ } from '../worldwide';
+import { addHandler, maybeInstrument, triggerHandlers } from './_handlers';
+
+const WINDOW = GLOBAL_OBJ as unknown as Window;
+
+export const SENTRY_XHR_DATA_KEY = '__sentry_xhr_v3__';
+
+/**
+ * Add an instrumentation handler for when an XHR request happens.
+ * The handler function is called once when the request starts and once when it ends,
+ * which can be identified by checking if it has an `endTimestamp`.
+ *
+ * Use at your own risk, this might break without changelog notice, only used internally.
+ * @hidden
+ */
+export function addXhrInstrumentationHandler(handler: (data: HandlerDataXhr) => void): void {
+  const type = 'xhr';
+  addHandler(type, handler);
+  maybeInstrument(type, instrumentXHR);
+}
+
+/** Exported only for tests. */
+export function instrumentXHR(): void {
+  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+  if (!(WINDOW as any).XMLHttpRequest) {
+    return;
+  }
+
+  const xhrproto = XMLHttpRequest.prototype;
+
+  fill(xhrproto, 'open', function (originalOpen: () => void): () => void {
+    return function (this: XMLHttpRequest & SentryWrappedXMLHttpRequest, ...args: any[]): void {
+      const startTimestamp = Date.now();
+
+      // open() should always be called with two or more arguments
+      // But to be on the safe side, we actually validate this and bail out if we don't have a method & url
+      const method = isString(args[0]) ? args[0].toUpperCase() : undefined;
+      const url = parseUrl(args[1]);
+
+      if (!method || !url) {
+        return;
+      }
+
+      this[SENTRY_XHR_DATA_KEY] = {
+        method,
+        url,
+        request_headers: {},
+      };
+
+      // if Sentry key appears in URL, don't capture it as a request
+      if (method === 'POST' && url.match(/sentry_key/)) {
+        this.__sentry_own_request__ = true;
+      }
+
+      const onreadystatechangeHandler: () => void = () => {
+        // For whatever reason, this is not the same instance here as from the outer method
+        const xhrInfo = this[SENTRY_XHR_DATA_KEY];
+
+        if (!xhrInfo) {
+          return;
+        }
+
+        if (this.readyState === 4) {
+          try {
+            // touching statusCode in some platforms throws
+            // an exception
+            xhrInfo.status_code = this.status;
+          } catch (e) {
+            /* do nothing */
+          }
+
+          const handlerData: HandlerDataXhr = {
+            args: [method, url],
```

Comment:
we used to just pass `args: args as [string, string]` here which is actually "wrong" or not really what is expected (I think), because `args` in here is the args passed to `send`, which is the body (that we capture separately above).